### PR TITLE
Multilevel Bayesian quadrature

### PR DESCRIPTION
**Pull request by tskarvone**
# In a Nutshell

This RP adds a basic multilevel Bayesian quadrature method that is described in Proposition 1 of https://arxiv.org/pdf/2210.08329.pdf.

# Detailed Description

- This is essentially nothing more than a loop that calls `bayesquad_from_data` a number of times and sums the outputs.
- For now the user is required to input the nodes and function evaluations, but it would be easy to include a version where a number of error tolerances that need to be met are provided instead. In that case the current method should probably be renamed e.g. `bayesquad_multileve_from_data`.
- Testing seems somewhat difficult. Only input handling and trivial cases where the multilevel method reduces to traditional BQ are tested.

It may be that this should be reviewed only after some or all of #746, #748 and #749 are merged.

# Related Issues

na

### CODE REVIEW

File: src/probnum/quad/_bayesquad.py [hunk 1]
```diff
@@ -284,6 +283,143 @@ def bayesquad_from_data(
...
+    \mu(x)` + \sum_{l=1}^L :math:`\int_\Omega [f_l(x) - f_{l-1}(x)] d \mu(x)`,
+
+    where :math:`f_l` is an increasingly accurate but also increasingly expensive
+    approximation to :math:`f`. Bayesian quadrature is subsequently applied to
+    independently infer each of the :math:`L+1` integrals and the outputs are summed
+    to infer :math:`\int_\Omega f(x) d \mu(x)`.
+
+    Parameters
+    ----------
+    nodes
```
Comment (mmahsereci) [inline src/probnum/quad/_bayesquad.py:310]: Are there any requirements? E.g., do the nodes need to be nested?

**Review by mmahsereci — COMMENTED** [review]
What about `multilevel_bayesquad_from_data` as name? It would reflect the paper title. 

In terms of tests, you can assert the instance of the return objects and the length of the info list.

File: src/probnum/quad/_bayesquad.py [hunk 2]
```diff
@@ -284,6 +283,143 @@ def bayesquad_from_data(
...
+    0.7252421350019139
+    """
+
+    if kernels is None:
+        kernels = (None,)
+    n_level = np.max([len(nodes), len(fun_diff_evals), len(kernels)])
+    if len(nodes) == 1:
+        nodes = n_level * (nodes[0],)
+    if len(kernels) == 1:
+        kernels = n_level * (kernels[0],)
```
[the remark this answers is not in the document]
Comment (tskarvone) [reply src/probnum/quad/_bayesquad.py:408]: The user is now required to input all kernels.

File: src/probnum/quad/_bayesquad.py (hunk 1 above)
Comment (tskarvone) [reply src/probnum/quad/_bayesquad.py:325]: No, there are no such requirements. The only requirement is that the number of nodes for a given level obviously needs to match the number of functions evaluations provided for that level.

File: src/probnum/quad/_bayesquad.py [hunk 3]
```diff
@@ -284,6 +283,143 @@ def bayesquad_from_data(
...
+    r"""Infer the value of an integral from given sets of nodes and function
+    evaluations using a multilevel method.
+
+    In multilevel Bayesian quadrature, the integral :math:`\int_\Omega f(x) d \mu(x)`
+    is (approximately) decomposed as a telescoping sum over :math:`L+1` levels:
+
+    .. math:: \int_\Omega f(x) d \mu(x) \approx :math:`\int_\Omega f_0(x) d
+    \mu(x)` + \sum_{l=1}^L :math:`\int_\Omega [f_l(x) - f_{l-1}(x)] d \mu(x)`,
+
+    where :math:`f_l` is an increasingly accurate but also increasingly expensive
```
[the remark this answers is not in the document]
Comment (tskarvone) [reply src/probnum/quad/_bayesquad.py:318]: I've tried to make it clearer that it is not necessary for $f_L$ to equal $f$.

**Comment by tskarvone** [comment]
> What about `multilevel_bayesquad_from_data` as name? It would reflect the paper title.

I've renamed the method as you suggest.

**Comment by tskarvone** [comment]
All the failed checks are about `src/probnum/quad/__init__.py::probnum.quad.multilevel_bayesquad_from_data` but I do not understand what the problem is.

File: tests/test_quad/test_bayesquad/test_bq.py [hunk 4]
```diff
@@ -219,3 +219,131 @@ def test_zero_function_gives_zero_variance_with_mle(rng):
...
+    assert len(infos) == n_level
+    # Only one kernel and one set of nodes
+    F, infos = multilevel_bayesquad_from_data(
+        nodes=nodes_1,
+        fun_diff_evals=fun_diff_evals_2,
+        kernels=kernels_1,
+        measure=measure,
+    )
+    assert isinstance(F, Normal)
+    assert len(infos) == n_level
```
Comment (mmahsereci) [inline tests/test_quad/test_bayesquad/test_bq.py:264]: Is this part of the test obsolete with the new changes?

File: tests/test_quad/test_bayesquad/test_bq.py [hunk 5]
```diff
@@ -219,3 +219,131 @@ def test_zero_function_gives_zero_variance_with_mle(rng):
...
+            measure=measure,
+        )
+    nodes_2 = (nodes_full[0], nodes_full[1])
+    with pytest.raises(ValueError):
+        _, _ = multilevel_bayesquad_from_data(
+            nodes=nodes_2,
+            fun_diff_evals=fun_diff_evals_2,
+            kernels=kernels_2,
+            measure=measure,
+        )
```
Comment (mmahsereci) [inline tests/test_quad/test_bayesquad/test_bq.py:281]: can you factor out the exception raises tests into a separate test such that one test function checks the output type and shape, and the other the exception raises based on wrong input?

File: src/probnum/quad/_bayesquad.py [hunk 6]
```diff
@@ -299,6 +298,138 @@ def bayesquad_from_data(
...
+    >>> fun_diff_evals = ()
+    >>> for l in range(n_level):
+    ...     n_l = 2*l + 1
+    ...     nodes += (np.reshape(np.linspace(0, 1, n_l), (n_l, input_dim)),)
+    ...     fun_diff_evals += (np.reshape(fun(nodes[l], l), (n_l,)),)
+    >>> F, infos = multilevel_bayesquad_from_data(nodes=nodes,
+    ...                                           fun_diff_evals=fun_diff_evals,
+    ...                                           domain=domain)
+    >>> print(F.mean)
+    0.7252421350019139
```
Comment (mmahsereci) [inline src/probnum/quad/_bayesquad.py:402]:
```suggestion
    >>> print(np.round(F.mean, 4))
    0.7252
```

this should fix the failing doctest.

**Review by mmahsereci — COMMENTED** [review]
hey Toni, some more comments, mainly on the tests, but they should be all fast to fix. 

The failing test was a doctest due to numerical imprecision. There is a comment on how to fix it. You can check doctests locally by running `tox -e py3 -- src/probnum/quad/` in the root directory.

File: src/probnum/quad/_bayesquad.py [hunk 7]
```diff
@@ -284,6 +283,143 @@ def bayesquad_from_data(
...
+    Parameters
+    ----------
+    nodes
+        *shape=(n_level,)* -- Tuple containing the locations for each level at which
+        the functionn evaluations are available as ``fun_diff_evals``. Each element
+        must be a shape=(n_eval, input_dim) ``np.ndarray``. If a tuple containing only
+        one element is provided, it is inferred that the same nodes ``nodes[0]`` are
+        used on every level.
+    fun_diff_evals
+        *shape=(n_level,)* --- Tuple containing the evaluations of :math:`f_l - f_{l-1}`
```
[the remark this answers is not in the document]
Comment (mmahsereci) [reply src/probnum/quad/_bayesquad.py:335]: Reopened as the comment doesn't seem to be resolved. A tuple does not have a `shape` property, but it does have a length (`len`) property :)

File: tests/test_quad/test_bayesquad/test_bq.py [hunk 8]
```diff
@@ -219,3 +219,131 @@ def test_zero_function_gives_zero_variance_with_mle(rng):
...
+        nodes_full += (measure.sample(n_sample=ns_1[l], rng=rng),)
+    F, infos = multilevel_bayesquad_from_data(
+        nodes=nodes_full,
+        fun_diff_evals=fun_diff_evals_1,
+        measure=measure,
+    )
+    assert isinstance(F, Normal)
+    assert len(infos) == n_level
+    # Only one set of nodes
+    kernels_1 = n_level * (kernel,)
```
[the remark this answers is not in the document]
Comment (tskarvone) [reply tests/test_quad/test_bayesquad/test_bq.py:239]: Fixed this. Hopefully the modification is a good way of doing this.

**Comment by tskarvone** [comment]
Thank you, Maren. That seems to have fixed most of the checks, though a Jupyter Notebook check still eludes me.

**Comment by mmahsereci** [comment]
> Thank you, Maren. That seems to have fixed most of the checks, though a Jupyter Notebook check still eludes me.

It doesn't seem to be related to the changes you made but a separate issue. 

I made some cosmetic changes. Can you check if you're OK with it?

**Comment by mmahsereci** [comment]
@JonathanWenger @marvinpfoertner As the jupyter test seems to be failing for other reasons (see #781 ) we'll merge this anyways if it's OK with everyone (once the rest of the tests pass of course.)

**Review by mmahsereci — APPROVED** [review]

**Comment by tskarvone** [comment]
Thanks again, Maren. Your changes look OK. Happy to have this merged now.

**Comment by JonathanWenger** [comment]
> @JonathanWenger @marvinpfoertner As the jupyter test seems to be failing for other reasons (see #781 ) we'll merge this anyways if it's OK with everyone (once the rest of the tests pass of course.)

I just fixed this, see #783 . But merging this PR first is also fine with me.

**Pull request merged**